Modify picker_jni to use different makeHit functions for different colliders

J0Nreynolds opened the pull request:
There are now two separate functions for creating GVRPickedObjects from picker_jni depending on the type of pick/collider.

Dependent on #1290.

@NolaDonato Please see if this accomplishes what you mentioned earlier.

liaxim commented:
I get conflicts when trying to merge this after merging https://github.com/Samsung/GearVRf/pull/1290. Suggest including the commit from https://github.com/Samsung/GearVRf/pull/1290 here. A commit you can remove after https://github.com/Samsung/GearVRf/pull/1290 gets merged.

J0Nreynolds commented:
Will do, thanks for catching that.

NolaDonato commented:
Nobody will use the arrays that you allocate in the constructor if mesh picking is enabled. Should we leave these null to avoid 3 extra allocations?

liaxim commented:
Needs rebase

J0Nreynolds commented:
@liaxim Rebased.

@NolaDonato I see your point that the allocations are now useless for PickingInformation that isn't generated for MeshColliders with coordinate picking enabled. However, before I make that change, I'd like to know how you think we should handle functions like getNormalX(), getNormalY(), etc. when the float array is null. Do we leave it up to the user to check the collider-type before querying these fields?

NolaDonato commented:
I personally think we should just remove these functions to avoid the issue. They can also use the existence of the arrays to determine whether there is meaningful information there or not.

J0Nreynolds commented:
The suggested changes have been made.

liaxim commented:
Needs update. Refers toGVRPickedObject.getHitX/Y/Z which are missing.

J0Nreynolds commented:
Rebased and updated old references to getHitX/Y/Z etc.

liaxim commented:
Since you removed getHitX/Y/Z could you please update the 3DCursorLibrary and the PickerTests? Thanks.

J0Nreynolds commented:
I've updated the 3D cursor library here and the picking tests [here.](https://github.com/gearvrf/GearVRf-Tests/pull/112)

liaxim commented:
The newly added canPickBoxCollider test keeps failing for me.

liaxim commented:
Please look into porting this change to the vulkan branch.

J0Nreynolds commented:
@liaxim Yes, sorry for not clarifying. I updated tests referring to getHitX/Y/Z, but also added a test for #1334. Before #1334, there was no isHit implementation for BoxColliders, so there's no way to pass that test without it.

liaxim commented:
Whenever I start using this pull request the picking tests tend to fail due to:
```
08-24 09:27:37.105  1519  1519 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
08-24 09:27:37.105  1519  1519 F DEBUG   : Build fingerprint: 'samsung/dreamqltesq/dreamqltesq:7.0/NRD90M/G950USQU1AQFJ:eng/test-keys'
08-24 09:27:37.105  1519  1519 F DEBUG   : Revision: '12'
08-24 09:27:37.106  1519  1519 F DEBUG   : ABI: 'arm'
08-24 09:27:37.106  1519  1519 F DEBUG   : pid: 1289, tid: 1496, name: GLThread 496  >>> org.gearvrf.tester <<<
08-24 09:27:37.106  1519  1519 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xd22337d0
08-24 09:27:37.106  1519  1519 F DEBUG   :     r0 bf800000  r1 0000000c  r2 d22337d0  r3 d0f46190
08-24 09:27:37.106  1519  1519 F DEBUG   :     r4 d0d78a54  r5 d21ecd54  r6 d0061c00  r7 d2197020
08-24 09:27:37.106  1519  1519 F DEBUG   :     r8 fffffffd  r9 fffffffc  sl fffffffe  fp d0d7880c
08-24 09:27:37.106  1519  1519 F DEBUG   :     ip d0d78818  sp d0d787d8  lr d330f90d  pc d330fb5c  cpsr 20070030
08-24 09:27:37.119  1519  1519 F DEBUG   : 
08-24 09:27:37.119  1519  1519 F DEBUG   : backtrace:
08-24 09:27:37.119  1519  1519 F DEBUG   :     #00 pc 0009cb5c  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr12MeshCollider5isHitERKNS_4MeshERKN3glm5tvec3IfLNS4_9precisionE0EEES9_b+335)
08-24 09:27:37.119  1519  1519 F DEBUG   :     #01 pc 0009cfb1  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr12MeshCollider5isHitERKN3glm5tvec3IfLNS1_9precisionE0EEES6_+296)
08-24 09:27:37.119  1519  1519 F DEBUG   :     #02 pc 000889bf  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (_ZN3gvr6Picker9pickSceneEPNS_5SceneERSt6vectorINS_12ColliderDataESaIS4_EEPNS_9TransformEffffff+194)
08-24 09:27:37.119  1519  1519 F DEBUG   :     #03 pc 00088fc9  /data/app/org.gearvrf.tester-1/lib/arm/libgvrf.so (Java_org_gearvrf_NativePicker_pickObjects+148)
08-24 09:27:37.119  1519  1519 F DEBUG   :     #04 pc 00567c6f  /data/app/org.gearvrf.tester-1/oat/arm/base.odex (offset 0x53d000)
```
Now it is likely not caused by this pull request but it just helps with the manifestation.

liaxim commented:
Crash traced to populateBarycentricCoords and colliderData.FaceIndex of -1.

liaxim commented:
Planning to merge on Friday. There will be a separate pull request for the PickerTests crashes.

liaxim commented:
@dnvasilev Can't cherry-pick/can't merge since the multiview commit is master only. Will let you resolve this.
